xapian-letor/api/featuremanager_internal.cc

```diff
@@ -0,0 +1,194 @@
+/** @file featuremanager_internal.cc
+ * @brief Internals of FeatureManager class
+ */
+/* Copyright (C) 2011 Parth Gupta
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include <xapian.h>
+
+#include "xapian-letor/featuremanager.h"
+#include "xapian-letor/featurevector.h"
+#include "xapian-letor/ranklist.h"
+#include "featuremanager_internal.h"
+
+#include <cstring>
+#include <cstdlib>
+#include <fstream>
+
+#include <map>
+#include <string>
+
+
+using namespace Xapian;
+using namespace std;
+
+
+std::string
+FeatureManager::Internal::getdid(const Document &doc) 
+{
+    string id="";
+    string data = doc.get_data();
+    string temp_id = data.substr(data.find("url=", 0), (data.find("sample=", 0) - data.find("url=", 0)));
+    id = temp_id.substr(temp_id.rfind('/') + 1, (temp_id.rfind('.') - temp_id.rfind('/') - 1));  //to parse the actual document name associated with the         documents if any
```

Comment:
It's worth wrapping these lines to try to get the bulk of them within 80 characters. (It's not always possible easily, but there's some straightforward approaches here that will work.)